src/convex/ConvexAvatarMultiToken.sol

```diff
@@ -0,0 +1,833 @@
+        external
+        onlyOwner
+        returns (bytes32 kekId)
+    {
+        if (_amountAsset == 0) {
+            revert NothingToDeposit();
+        }
+
+        address vaultAddr = privateVaults[_pid];
+        if (vaultAddr == address(0)) {
+            revert NoPrivateVaultForPid(_pid);
+        }
+
+        IStakingProxy proxy = IStakingProxy(vaultAddr);
+        address stakingToken = proxy.stakingToken();
+        IFraxUnifiedFarm farm = IFraxUnifiedFarm(proxy.stakingAddress());
+
+        IERC20MetadataUpgradeable(stakingToken).safeTransferFrom(msg.sender, address(this), _amountAsset);
+
+        if (_additionalFunds) {
+            kekId = kekIds[vaultAddr];
+            if (kekId == bytes32(0)) {
+                revert NoExistingLockInPrivateVault(vaultAddr);
+            }
+
+            proxy.lockAdditional(kekId, _amountAsset);
+        } else {
+            // NOTE: we always try to lock for the min duration allowed
+            uint256 minLockTime = farm.lock_time_min();
+            if (minLockTime > MAX_LOCKING_TIME) {
+                revert MinLockTimeMisconfigured(address(farm), minLockTime);
+            }
+            kekId = proxy.stakeLocked(_amountAsset, farm.lock_time_min());
+            /// @dev detailed required to enable withdrawls later
+            kekIds[vaultAddr] = kekId;
+        }
+
+        emit Deposit(stakingToken, _amountAsset, block.timestamp);
+    }
+
+    /// @notice Unstakes all staked assets and transfers them back to owner. Can only be called by owner or manager.
+    /// @dev Internally in the unified farm contract calls `getReward` collecting rewards and updating the balances
+    /// @param _pid Pid target to withdraw from avatar private vault
+    function withdrawFromPrivateVault(uint256 _pid) external onlyOwnerOrManager {
+        address vaultAddr = privateVaults[_pid];
+
+        if (vaultAddr == address(0)) {
+            revert NoPrivateVaultForPid(_pid);
+        }
+
+        IStakingProxy proxy = IStakingProxy(vaultAddr);
+        IFraxUnifiedFarm unifiedFarm = IFraxUnifiedFarm(proxy.stakingAddress());
+
+        uint256 lockedBal = unifiedFarm.lockedLiquidityOf(vaultAddr);
+        if (lockedBal == 0) {
+            revert NothingToWithdraw();
+        }
+
+        bytes32 kek = kekIds[vaultAddr];
+
+        /// @dev: only inspecting index `0` given that we only use an unique lock id per private vault
+        (,,, uint256 endingTimestamp,) = unifiedFarm.lockedStakes(vaultAddr, 0);
+
+        if (block.timestamp < endingTimestamp) {
+            revert NoExpiredLock(vaultAddr, kek);
+        }
+
+        proxy.withdrawLockedAndUnwrap(kek);
+
+        IERC20MetadataUpgradeable curveLp = IERC20MetadataUpgradeable(proxy.curveLpToken());
+        uint256 curveLpBalance = curveLp.balanceOf(address(this));
+        curveLp.safeTransfer(owner(), curveLpBalance);
+
+        delete kekIds[vaultAddr];
```

Comment:
Per discussions we've already had, this is not CEI conformant (change storage after external interactions)

This function is callable only by owner / manager so it's fine.

But you should fix it to avoid making the mistake when stakes are higher, recommend you read from storage (`kek`) then just delete the value from storage right after (before the external calls)

Reply:
added suggestion in  c376d84c41c74160ef5d845e12eb9bc06a50ac0c